### PR TITLE
Multiple models support, reprise

### PR DESCRIPTION
**Pull request by cchudant**
## Description

Built on top of #50 
Commit history may need tidying up (& rebased onto master)

## Related Issue

#39 

## Type of change

- [x] This change requires a documentation update
- [x] This change affects the client
- [x] This change affects the server
- [x] This change affects the API
- [x] This change only concerns the documentation
- [x] **This change is backward incompatible**

## How Has This Been Tested?

Added integration test & updated python tests

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have updated the documentation according to my changes

### CODE REVIEW

**Comment by JoFrost** [comment]
Seeing the kind of changes, I don't think it's necessary, but @clauverjat might still want to take a look.
By the way, can you please fix the conflicts issues?

**Comment by cchudant** [comment]
@JoFrost issues are fixed and i squashed my "to squash" git commits :sweat_smile:

File: server/blindai_sgx/src/model_store.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+use crate::model::{InferenceModel, ModelDatumType, OnnxModel};
+
+struct InnerModelStore {
+    models_by_id: HashMap<Uuid, InferenceModel>,
+    onnx_by_hash: HashMap<Vec<u8>, (usize, Arc<OnnxModel>)>,
+}
+
+/// This is where model are stored.
+pub struct ModelStore {
+    inner: RwLock<InnerModelStore>,
```
Comment (JoFrost) [inline server/blindai_sgx/src/model_store.rs:23]: Cool. Let's be careful about lock poisoning, though.

**Review by JoFrost — APPROVED** [review]
Ok. Careful with the Rwlock though because of the lock poisoning, but I did not notice any unchecked unwrap after getting the lock, so if it happens, it should be either very rare or for a good reason.

File: server/blindai_app/src/main.rs [hunk 2]
```diff
@@ -74,11 +74,11 @@ impl untrusted_local_app_server::UntrustedLocalApp for State {
     }
 }
 
-fn fill_blank_and_print(content: &str, size: usize)
-{
+fn fill_blank_and_print(content: &str, size: usize) {
     let trail_char = "#";
     let trail: String = trail_char.repeat((size - 2 - content.len()) / 2);
-    let trail2: String = trail_char.repeat(((size - 2 - content.len()) as f32 / 2.0).ceil() as usize);
+    let trail2: String =
```
Comment (clauverjat) [inline server/blindai_app/src/main.rs:80]: Assuming perfect arithmetic (no precision loss due to floating point numbers) (size - 2 - content.len()) as f32 / 2.0).ceil() as usize is actually the same as (size - 1 - content.len())/2  

todo: replace with the integer only formula

File: server/blindai_sgx/src/model_store.rs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+            // HashMap entry api requires only one lookup and should be prefered than .get()
+            // followed with .insert()
+
+            // deduplication support
+            let model = match models.onnx_by_hash.entry(model_hash_vec.clone()) {
+                Entry::Occupied(mut entry) => {
+                    let (num, onnx) = entry.get_mut();
+                    *num += 1;
+                    info!("Reusing an existing ONNX entry for model. (n = {})", *num);
```
Comment (clauverjat) [inline server/blindai_sgx/src/model_store.rs:62]: This feature might be useful for users, but beware if there are multiple model uploaders. You actually leak what models others have previously updated...

**Review by clauverjat — COMMENTED** [review]
Great job overall. I put a few reviews. 

Note that I just did a quick review of the code. I did not do a security review

**Pull request merged**